File: demos/pax-britannica/pax-britannica/src/de/swagner/paxbritannica/Targeting.java
package de.swagner.paxbritannica;

import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;

import de.swagner.paxbritannica.bomber.Bomber;
import de.swagner.paxbritannica.factory.FactoryProduction;
import de.swagner.paxbritannica.fighter.Fighter;
import de.swagner.paxbritannica.frigate.Frigate;

public class Targeting {

	public static boolean onScreen(Vector2 position) {
		return onScreen(position, 0);
	}

	public static boolean onScreen(Vector2 position, float buffer) {
		return position.x >= Constants.screenLeft - buffer && position.x <= Constants.screenRight + buffer && position.y >= Constants.screenBottom - buffer
				&& position.y <= Constants.screenTop + buffer;
	}

	/*
	 * returns the closest target of the given type 0 = Fighter 1 = Bomber 2 =
	 * Frigate 3 = Factory
	 */
	public static Ship getNearestOfType(Ship source, int shipType) {
		if (shipType == 0)
			return getNearestOfType(source, GameInstance.getInstance().fighters);
		else if (shipType == 3)
			return getFactoryWithHighestHealth(source, GameInstance.getInstance().factorys);
		else if (shipType == 1)
			return getNearestOfType(source, GameInstance.getInstance().bombers);
		else if (shipType == 2)
			return getNearestOfType(source, GameInstance.getInstance().frigates);
		else
			return null;
	}

	private static Ship getFactoryWithHighestHealth(Ship source, Array<Ship> ships) {
		// find the closest one!
		Ship closestShip = null;
		float highestHealth = Float.MIN_VALUE;

		for (int i = 0; i < ships.size; i++) {
			Ship ship = ships.get(i);
			float currentHealth = ship.hitPoints+(((FactoryProduction)ship).harvestRate*500);

			if (ship.alive && source.id != ship.id && onScreen(ship.collisionCenter) && (currentHealth > highestHealth)) {
				closestShip = ship;
				highestHealth = currentHealth;
			}
		}

		return closestShip;
	}
	
	private static Ship getNearestOfType(Ship source, Array<Ship> ships) {
		// find the closest one!
		Ship closestShip = null;
		float closestDistanze = Float.MAX_VALUE;

		for (int i = 0; i < ships.size; i++) {
			Ship ship = ships.get(i);
			float currentDistance = source.collisionCenter.dst(ship.collisionCenter);

			if (ship.alive && source.id != ship.id && onScreen(ship.collisionCenter) && (currentDistance < closestDistanze)) {
				//skip if ship is not targeting source ship
				if(ship instanceof Fighter) {
					if(((Fighter) ship).ai.target!=null && ((Fighter) ship).ai.target.id != source.id) {
						continue;
					}
				}
				if(ship instanceof Bomber) {
					if(((Bomber) ship).ai.target!=null && ((Bomber) ship).ai.target.id != source.id) {
						continue;
					}
				}
				if(ship instanceof Frigate) {
					if(((Frigate) ship).ai.target!=null && ((Frigate) ship).ai.target.id != source.id) {
						continue;
					}
				}
				closestShip = ship;
				closestDistanze = currentDistance;
			}
		}

		return closestShip;
	}

	/*
	 * return a random ship of the desired type that's in range
	 * 0 = Fighter 1 = Bomber 2 = Frigate 3 = Factory
	 */
	public static Ship getTypeInRange(Ship source, int shipType, float range) {
		if (shipType == 0)
			return getTypeInRange(source, GameInstance.getInstance().fighters, range);
		else if (shipType == 3)
			return getTypeInRange(source, GameInstance.getInstance().factorys, range);
		else if (shipType == 1)
			return getTypeInRange(source, GameInstance.getInstance().bombers, range);
		else if (shipType == 2)
			return getTypeInRange(source, GameInstance.getInstance().frigates, range);
		else
			return null;
	}

	/**
	 * return a random ship of the desired type that's in range
	 * @param source
	 * @param ships
	 * @param range
	 * @return
	 */
	private static Ship getTypeInRange(Ship source, Array<Ship> ships, float range) {
		Array<Ship> shipsInRange = new Array<Ship>();
		float range_squared = range * range;

		for (int i = 0; i < ships.size; i++) {
			Ship ship = ships.get(i);
			float currentDistance = source.collisionCenter.dst(ship.collisionCenter);

			if (ship.alive && source.id != ship.id && onScreen(ship.collisionCenter) && (currentDistance < range_squared)) {
				shipsInRange.add(ship);
			}
		}

		if (shipsInRange.size > 0) {
			return shipsInRange.get(MathUtils.random(0, shipsInRange.size - 1));
		} else {
			return null;
		}
	}
}

File: demos/pax-britannica/pax-britannica-android/assets/data/components/targeting.lua
local v2 = require 'dokidoki.v2'

function on_screen(pos)
  return pos.x >= game.constants.screen_left and
         pos.x <= game.constants.screen_right and
         pos.y >= game.constants.screen_bottom and
         pos.y <= game.constants.screen_top
end

-- returns the closest target of the given type
function get_nearest_of_type(source, ship_type)
  local ships = game.actors.get(ship_type)

  -- find the closest one!
  local closest_ship
  local closest_square_magnitude

  for _, ship in ipairs(ships) do
    local square_magnitude = v2.sqrmag(ship.transform.pos - source.transform.pos)
    
    if source.ship.player ~= ship.ship.player and on_screen(ship.transform.pos) and (not closest_ship or square_magnitude < closest_square_magnitude) then
      closest_ship = ship
      closest_square_magnitude = square_magnitude
    end
  end

  return closest_ship
end

-- return a random ship of the desired type that's in range
function get_type_in_range(source, ship_type, range)
  local ships = game.actors.get(ship_type)
  local ships_in_range = {}
  local range_squared = range * range

  for _, ship in ipairs(ships) do
    local square_magnitude = v2.sqrmag(ship.transform.pos - source.transform.pos)
    
    if source.ship.player ~= ship.ship.player and on_screen(ship.transform.pos) and (square_magnitude < range_squared) then
      ships_in_range[#ships_in_range+1] = ship
    end
  end

  if #ships_in_range > 0 then
    return ships_in_range[math.random(1, #ships_in_range)]
  else
    return false
  end
end
